extension/maxminddb.c

```diff
@@ -166,6 +171,106 @@ static PyObject *Reader_get(PyObject *self, PyObject *args)
     return py_obj;
 }
 
+static int ip_converter(PyObject *obj, struct sockaddr_storage *ip_address)
+{
+#if PY_MAJOR_VERSION >= 3
+    if (PyUnicode_Check(obj)) {
+        Py_ssize_t len;
+        const char *ipstr = PyUnicode_AsUTF8AndSize(obj, &len);
+# else
+    if (PyUnicode_Check(obj) || PyString_Check(obj)) {
+        // Although this should work on Python 3, we will hopefully delete
+        // this soon and the Python 3 version is cleaner.
+        const char *ipstr = PyString_AsString(obj);
+        Py_ssize_t len = strlen(ipstr);
+#endif
+        if (!ipstr) {
+            PyErr_SetString(PyExc_TypeError,
+                            "argument 1 contains an invalid string");
+            return 0;
+        }
+        if (strlen(ipstr) != (size_t)len) {
+            PyErr_SetString(PyExc_TypeError,
+                            "argument 1 contains an embedded null character");
+            return 0;
+        }
+
+        struct addrinfo hints = {
+            .ai_family   = AF_UNSPEC,
+            .ai_flags    = AI_NUMERICHOST,
+            // We set ai_socktype so that we only get one result back
+            .ai_socktype = SOCK_STREAM
+        };
+
+        struct addrinfo *addresses = NULL;
+        int gai_status = getaddrinfo(ipstr, NULL, &hints, &addresses);
+        if (gai_status) {
+            PyErr_Format(PyExc_ValueError,
+                         "'%s' does not appear to be an IPv4 or IPv6 address.",
+                         ipstr);
+            return 0;
+        }
+        memcpy(ip_address, addresses->ai_addr, addresses->ai_addrlen);
```

Comment:
I wonder if we should check if addresses is null. I don't know how that would be possible though.

Reply:
Hmm, I think `getaddrinfo` would have to be out of spec for that to happen, but I suppose it doesn't hurt.